torch/custom_class.h

```diff
@@ -207,16 +224,19 @@ class class_ {
         getstate_schema.returns().size() == 1,
         "__getstate__ should return exactly one value for serialization. Got: ",
         format_getstate_schema());
-    auto ser_type = getstate_schema.returns().at(0).type();
-    auto setstate_schema = classTypePtr->getMethod("__setstate__").getSchema();
-    auto arg_type = setstate_schema.arguments().at(1).type();
-    TORCH_CHECK(
-        (*arg_type == *ser_type),
-        "__setstate__'s argument should be the same type as the "
-        "return value of __getstate__. Got ",
-        arg_type->repr_str(),
-        " but expected ",
-        ser_type->repr_str());
+
+    if (validate_setstate_input_type) {
```

Comment:
(moving over @dzhulgakov comment from https://github.com/pytorch/pytorch/pull/43230)

if we want to allow IValue setter officially, we could also not have a separate split between def_pickle and def_pickle_do_not_use, just have a single one and allow the setstate argument to be Any. You can be even fancier and check isSubType on the type to just relax condition (i.e. that the setstate's argument type should be same or broader than the getstate's). I think it's pretty clean

I actually don't see a big harm in allowing the set

Reply:
Sure.  One other thing to think about is whether we are (1) checking this in for the long term or (2) as a hack to be deleted ASAP after VIP customers migrate.  Currently it's following the second, but we can switch to the first if that's the consensus.

Reply:
Let's go for long term (in which case we should modify the docs too). The right thing then is to change `==` into `isSubTypeOf` below. It'd work for IValue (which has type `Any`) but would also work for some more narrow relaxations. Seems like our offline discussion converged to that too, let's do it